entity_gym/entity_gym/environment.py

```diff
@@ -388,4 +506,205 @@ def act(
                 observations.append(new_obs)
             else:
                 observations.append(obs)
-        return batch_obs(observations)
+        return batch_obs(observations, self.cls.obs_space(), self.cls.action_space())
+
+
+class CloudpickleWrapper:
+    def __init__(self, var: Any):
+        self.var = var
+
+    def __getstate__(self) -> Any:
+        return cloudpickle.dumps(self.var)
+
+    def __setstate__(self, var: Any) -> None:
+        self.var = cloudpickle.loads(var)
+
+
+class MsgpackConnectionWrapper(object):
+    """
+    Use msgpack instead of pickle to send and recieve data from workers.
+    """
+
+    def __init__(self, conn: Connection) -> None:
+        self._conn = conn
+
+    def close(self) -> None:
+        self._conn.close()
+
+    def send(self, data: Any) -> None:
+        s = msgpack_numpy.dumps(
+            data, default=MsgpackConnectionWrapper.ragged_buffer_encode
+        )
+        self._conn.send_bytes(s)
+
+    def recv(self) -> Any:
+        data_bytes = self._conn.recv_bytes()
+        return msgpack_numpy.loads(
+            data_bytes,
+            object_hook=MsgpackConnectionWrapper.ragged_buffer_decode,
+            strict_map_key=False,
+        )
+
+    @classmethod
+    def ragged_buffer_encode(cls, obj: Any) -> Any:
+        if isinstance(obj, RaggedBufferF32) or isinstance(obj, RaggedBufferI64) or isinstance(obj, RaggedBufferBool):  # type: ignore
+            flattened = obj.as_array()
+            lengths = obj.size1()
+            return {
+                "__flattened__": msgpack_numpy.encode(flattened),
+                "__lengths__": msgpack_numpy.encode(lengths),
+            }
+        elif hasattr(obj, "__dict__"):
+            return {"__classname__": obj.__class__.__name__, "data": vars(obj)}
+        else:
+            return obj
+
+    @classmethod
+    def ragged_buffer_decode(cls, obj: Any) -> Any:
+        if "__flattened__" in obj:
+            flattened = msgpack_numpy.decode(obj["__flattened__"])
+            lengths = msgpack_numpy.decode(obj["__lengths__"])
+
+            dtype = flattened.dtype
+
+            if dtype == np.float32:
+                return RaggedBufferF32.from_flattened(flattened, lengths)
+            elif dtype == int:
+                return RaggedBufferI64.from_flattened(flattened, lengths)
+        elif "__classname__" in obj:
+            cls_name = globals()[obj["__classname__"]]
+            return cls_name(**obj["data"])
```

Comment:
Why do we need this? I assume to serialize dataclasses? This probably allows for arbitrary code execution, if we need this we should probably at least have a whitelist of allowed classes.

Reply:
Yes its so we can serialize classes like ObsBatch etc. I'm not sure if it does allow for arbitrary execution. My understanding is that `globals()` only allows classes that are in scope, so the Whitelist is essentially already included in the imports of the file.

Reply:
Well, it would be very easy to accidentally add an import that allows for something destructive, so I'd sleep better with an explicit whitelist. Globals has access to things like `__builtins__` and `__loader__`, which I've not found an obvious way to utilize, but I wouldn't be completely certain that it's impossible to get access to something like `eval`, particularly over multiple iterations.

```
>>> globals()
{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>}
```

Reply:
Sounds good, I agree its better to be defensive here. I'm no expert on python security, so whitelisting sounds smart.